src/eth.rs

```diff
@@ -173,14 +308,180 @@ impl VitalikProvider {
 
         Ok(deal_info)
     }
+}
 
-    // the validator should be able to handle if proofs get sent twice on accident
-    // return the block number that the proof made it into.
-    pub async fn post_proof(&self, _deal_id: &DealID, _proof: Proof) -> Result<BlockNum> {
-        unimplemented!("write me :)")
+/// A builder for the EthClient
+pub struct EthClientBuilder {
+    /// The URL of the Ethereum API to use. Required to build the client.
+    api_url: String,
+    /// The API Key to use for the Ethereum API. Required to build the client.
+    api_key: String,
+    /// The (optional) chain ID to use for the Ethereum API. Defaults to 1 (mainnet).
+    chain_id: Option<u64>,
+    /// The (optional) private key to use for signing payable transactions. Not required for read-only operations.
+    private_key: Option<String>,
+    /// The (optional) Contract Address to use for the EthClient. Not required for read-only
+    contract_address: Option<Address>,
+    /// The (optional) timeout to use for all Ethereum interactions
+    timeout: Option<u64>,
+}
+
+/// Default EthClientBuilder
+impl Default for EthClientBuilder {
+    /// Initialize an EthClientBuilder from the environment
+    /// Falls back to defaults if no environment variables are set
+    fn default() -> Self {
+        dbg!("Initializing EthClientBuilder from environment");
+        // Read the Api Url from the environment. Default to the mainnet Infura API
+        let api_url = env::var("ETH_API_URL")
+            .unwrap_or("https://mainnet.infura.io/v3/".to_string());
+        dbg!(format!("  API_URL: {}", &api_url));
+        // Read the Api Key from the environment. Raise an error if it is not set
+        let api_key = env::var("ETH_API_KEY").expect("ETH_API_KEY must be set");
+        dbg!(format!("  API_KEY: {}", &api_key));
+        // Try and Read the Chain ID from the environment. Default to 1 (mainnet)
+        let chain_id = env::var("ETH_CHAIN_ID")
+            .unwrap_or("1".to_string()).parse::<u64>().ok();
+        dbg!(format!("  CHAIN_ID: {:?}", &chain_id));
+        // Try and Read the Private Key from the environment. Default to None
+        let private_key = match env::var("ETH_PRIVATE_KEY").ok() {
+            Some(key) => Some(key),
+            None => None,
+        };
+        dbg!(format!("  PRIVATE_KEY: {:?}", &private_key));
+        // Try and Read the Contract Address from the environment. Default to None if not set
+        let contract_address = match env::var("ETH_CONTRACT_ADDRESS").ok() {
+            Some(address) => Some(address.parse::<Address>().unwrap()),
+            None => None,
+        };
+        dbg!(format!("  CONTRACT_ADDRESS: {:?}", &contract_address));
+
+        // TODO: Add param for ABI
+        Self {
+            api_url,
```

Comment:
this just feels really really unnecessary

Reply:
like this whole builder situation, i'm gonna veto sorry

Reply:
it reduces code bloat

Reply:
its gone